exporter/loadbalancingexporter/resolver.go

```diff
@@ -0,0 +1,33 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package loadbalancingexporter
+
+import "context"
+
+// resolver determines the contract for sources of backend endpoint information
+type resolver interface {
+	// resolve returns the current list of endpoints.
+	// returns either a non-nil error and a nil list of endpoints, or a non-nil list of endpoints and nil error.
+	resolve(context.Context) ([]string, error)
+
+	// start signals the resolver to start its work
+	start(context.Context) error
+
+	// shutdown signals the resolver to finish its work
+	shutdown(context.Context) error
```

Comment:
Does this provide any guarantees that callbacks will not be called after this call returns or it is undefined behavior?

Reply:
Great question, I added a comment here with the expected behavior, but in short: yes. It's currently only relevant for the DNS resolver, and here's what it does:

1. while a resolution is happening, a wait group is incremented
1. once the resolution finishes, the wait group is decreased
1. if a shutdown happens, a message is sent to the stop channel, which is in the same select as the resolve function, so, it will not happen concurrently to the steps above
1. once the periodic resolution gets the stop message, it returns
1. the shutdown function will send a message to the channel and wait for the wait group to be completed

In short: a client might call shutdown, and it will block until the current resolution has finished. During this block, it might happen that the resolution triggers the `onChange` callbacks, but it will not trigger them once shutdown returns.